Rm unused variables and equations

SarahAlidoost opened the pull request:
### Description
Please add a description of the changes proposed in the pull request.

### Checklist
- [ ] Add a reference to related issues.
- [ ] @mentions of the person or team responsible for reviewing proposed changes.
- [ ] This pull request has a descriptive title.
- [ ] Code is written according to the [guidelines](http://cnl.sogang.ac.kr/cnlab/lectures/programming/matlab/Richard_Johnson-MatlabStyle2_book.pdf).
- [ ] The checks by [MISS_HIT style checker and linter](https://github.com/EcoExtreML/STEMMUS_SCOPE/blob/main/CONTRIBUTING.md#follow-matlab-style-guidelines), below the pull request, are successful (green).
- [ ] Documentation is available.
- [ ] Add changes to the [changelog file](CHANGELOG.md) under section `Unreleased`.
- [ ] Model runs successfully.
- [ ] Ask a meinatainer to re-generate exe file if matlab codes are changed. About how to create an exe file, see [exe readme](https://github.com/EcoExtreML/STEMMUS_SCOPE/blob/main/run_model_on_snellius/exe/README.md).

SarahAlidoost commented:
@MostafaGomaa93 in this PR, unused variables and equations are commented out, only `botmSoilLevel` is removed. see [Files changed](https://github.com/EcoExtreML/STEMMUS_SCOPE/pull/239/files).

@BSchilperoort the exe file is re-genertaed. please let me know if more fixes are needed.

MostafaGomaa93 commented on src/+groundwater/readGroundwaterSettings.m:
```diff
@@ -94,32 +86,32 @@ soilThick           cumulative soil layers thickness (from top to bottom)
     end
 
-    indxBotmLayer_R = indxBotmLayer_R; % index of bottom layer that contains current headBotmLayer
+    % indxBotmLayer_R: index of bottom layer that contains current headBotmLayer
```
I think we can remove this line at all. It was my mistake that I didn't remove it

BSchilperoort commented:
I found it difficult to understand the current code.


Why not reduce the initialization to:
```matlab
function GroundwaterSettings = readGroundwaterSettings()
    % Activate/deactivate Groundwater coupling
    GroundwaterSettings.GroundwaterCoupling = 0; % (value = 0 -> deactivate coupling, or = 1 -> activate coupling); default = 0, update value to = 1 -> through BMI

    % Initialize the variables (head, temperature, air pressure) at the bottom boundary (start of saturated zone)
    GroundwaterSettings.tempBotm = 0; % groundwater temperature (C), received from MODFLOW through BMI

    % Call MODFLOW layers information (number of aquifer layers and their elevations, etc)
    GroundwaterSettings.gw_Dep = 1.0; % elevation of the top surface aquifer layer
    %     to avoid model crashing, assign minimum value of 1 cm
end
```

Move the soil thickness calculation to a more appropriate place. It does nothing with the groundwater inputs:
```matlab
    % Calculate soil layers thickness (cumulative layers thickness; e.g. 1, 2, 3, 5, 10, ......., 480, total soil depth)
    soilThick = zeros(NN, 1); % cumulative soil layers thickness
    soilThick(1) = 0;
    DeltZ = ModelSettings.DeltZ;
    DeltZ_R = ModelSettings.DeltZ_R;

    % Load model settings
    ModelSettings = io.getModelSettings();
    NN = ModelSettings.NN; % Number of nodes
    NL = ModelSettings.NL; % Number of layers

    for i = 2:NL
        soilThick(i) = soilThick(i - 1) + DeltZ_R(i - 1);
    end
    soilThick(NN) = ModelSettings.Tot_Depth; % total soil depth
```

<hr>

Also, the `indxBotmLayer` thing is a bit problematic. It is calculated in `readGroundwaterSettings.m` based on a groundwater depth value, but this is only run once at initialization, even though it's supposed to skip saturated layers:

```matlab
        % index of bottom layer after neglecting saturated layers (from bottom to top)
        indxBotm = GroundwaterSettings.indxBotmLayer;
```

So, shouldn't the following code be run every time step...?

```matlab
    for i = 1:NL
        midThick = (soilThick(i) + soilThick(i + 1)) / 2;
        if gw_Dep >= soilThick(i) && gw_Dep < soilThick(i + 1)
            if gw_Dep < midThick
                indxBotmLayer_R = i;
            elseif gw_Dep >= midThick
                indxBotmLayer_R = i + 1;
            end
            break
        elseif gw_Dep >= soilThick(i + 1)
            continue
        end
    end
```

<hr>

If these things are fixed, then the BMI implementation will be simple: only three variables (coupling toggle, temperature of the groundwater, groundwater head at top aquifer) have to be implemented.

MostafaGomaa93 commented:
> I found it difficult to understand the current code.
> 
> Why not reduce the initialization to:
> 
> ```matlab
> function GroundwaterSettings = readGroundwaterSettings()
>     % Activate/deactivate Groundwater coupling
>     GroundwaterSettings.GroundwaterCoupling = 0; % (value = 0 -> deactivate coupling, or = 1 -> activate coupling); default = 0, update value to = 1 -> through BMI
> 
>     % Initialize the variables (head, temperature, air pressure) at the bottom boundary (start of saturated zone)
>     GroundwaterSettings.tempBotm = 0; % groundwater temperature (C), received from MODFLOW through BMI
> 
>     % Call MODFLOW layers information (number of aquifer layers and their elevations, etc)
>     GroundwaterSettings.gw_Dep = 1.0; % elevation of the top surface aquifer layer
>     %     to avoid model crashing, assign minimum value of 1 cm
> end
> ```
> 
> Move the soil thickness calculation to a more appropriate place. It does nothing with the groundwater inputs:
> 
> ```matlab
>     % Calculate soil layers thickness (cumulative layers thickness; e.g. 1, 2, 3, 5, 10, ......., 480, total soil depth)
>     soilThick = zeros(NN, 1); % cumulative soil layers thickness
>     soilThick(1) = 0;
>     DeltZ = ModelSettings.DeltZ;
>     DeltZ_R = ModelSettings.DeltZ_R;
> 
>     % Load model settings
>     ModelSettings = io.getModelSettings();
>     NN = ModelSettings.NN; % Number of nodes
>     NL = ModelSettings.NL; % Number of layers
> 
>     for i = 2:NL
>         soilThick(i) = soilThick(i - 1) + DeltZ_R(i - 1);
>     end
>     soilThick(NN) = ModelSettings.Tot_Depth; % total soil depth
> ```
> 
> Also, the `indxBotmLayer` thing is a bit problematic. It is calculated in `readGroundwaterSettings.m` based on a groundwater depth value, but this is only run once at initialization, even though it's supposed to skip saturated layers:
> 
> ```matlab
>         % index of bottom layer after neglecting saturated layers (from bottom to top)
>         indxBotm = GroundwaterSettings.indxBotmLayer;
> ```
> 
> So, shouldn't the following code be run _every time step_...?
> 
> ```matlab
>     for i = 1:NL
>         midThick = (soilThick(i) + soilThick(i + 1)) / 2;
>         if gw_Dep >= soilThick(i) && gw_Dep < soilThick(i + 1)
>             if gw_Dep < midThick
>                 indxBotmLayer_R = i;
>             elseif gw_Dep >= midThick
>                 indxBotmLayer_R = i + 1;
>             end
>             break
>         elseif gw_Dep >= soilThick(i + 1)
>             continue
>         end
>     end
> ```
> 
> If these things are fixed, then the BMI implementation will be simple: only three variables (coupling toggle, temperature of the groundwater, groundwater head at top aquifer) have to be implemented.

Hi Bart, its fine for me to split this function into two as you suggested if that will increase the readability of the code

MostafaGomaa93 commented:
> So, shouldn't the following code be run _every time step_...?
> 
> ```matlab
>     for i = 1:NL
>         midThick = (soilThick(i) + soilThick(i + 1)) / 2;
>         if gw_Dep >= soilThick(i) && gw_Dep < soilThick(i + 1)
>             if gw_Dep < midThick
>                 indxBotmLayer_R = i;
>             elseif gw_Dep >= midThick
>                 indxBotmLayer_R = i + 1;
>             end
>             break
>         elseif gw_Dep >= soilThick(i + 1)
>             continue
>         end
>     end
> ```
> 

Yes, `gw_Dep` has to be calculated everytime step based on the updated `GroundwaterSettings.headBotmLayer`

MostafaGomaa93 commented:
> @MostafaGomaa93 can you check the calculation of `indxBotmLayer` and if it should be calculated every time step?

Yes, `indxBotmLayer` has to be updated everytime step of MODFLOW

SarahAlidoost commented:
@MostafaGomaa93 can you review the changes in this pull request? Please feel free to suggest better names for functions/variables, if needed. and also double-check the logic in this [code block](https://github.com/EcoExtreML/STEMMUS_SCOPE/blob/be505be537def40171797d02c7292f91a09d24bd/src/STEMMUS_SCOPE.m#L335-L352). Thanks.

MostafaGomaa93 commented on src/+groundwater/updateRunoff.m:
Could we change the file name to updateDunnianRunoff ?

if dunnianRunoff is not zero, Precip_msr will change and need to be updated. I adjust the code accordingly

MostafaGomaa93 commented on src/+energy/calculateBoundaryConditions.m:
The groundwater temperature data in many situations is not there, so I made a slight change to account for such situation

MostafaGomaa93 commented on src/+energy/solveEnergyBalanceEquations.m:
Same as for the above change. The groundwater temperature data in many situations is not there, so I made a slight change to account for such situation

BSchilperoort replied:
This won't work with well the BMI interface. It's not good to leave it undefined.
Alternatively, you can use a value to denote it's not available, e.g. `NaN`.

MostafaGomaa93 replied:
> This won't work with well the BMI interface. It's not good to leave it undefined. Alternatively, you can use a value to denote it's not available, e.g. `NaN`.

The `GroundwaterSettings.tempBotm` is not undefined. I just add the if condition in case the`GroundwaterSettings.tempBotm` in this [line](https://github.com/EcoExtreML/STEMMUS_SCOPE/blob/8f9699e9b5a480166345d07ff6b3cb35938e12f2/src/%2Bgroundwater/readGroundwaterSettings.m#L28) = NaN. Or maybe I mis-understand you

MostafaGomaa93 replied:
Or do you mean i need to change the if condition from `if isempty(GroundwaterSettings.tempBotm)` to  `if GroundwaterSettings.tempBotm == NaN` ?

BSchilperoort replied:
Yes.

`tempBotm` is not an array so `isempty` is not a good check.

The default value of `tempBotm` can be NaN. If the temperature is available, the user can set it through the BMI.

MostafaGomaa93 replied:
Okay, i changed `isempty` to `isnan`

BSchilperoort commented on src/+groundwater/initializeGroundwaterSettings.m:
```diff
@@ -0,0 +1,41 @@
+
+    % Initialize the variables (head, temperature) at the bottom boundary (start of saturated zone)
+    GroundwaterSettings.headBotmLayer = 1950.0; % groundwater head (cm) at bottom layer
+    GroundwaterSettings.tempBotm = 17.0; % groundwater temperature at bottom layer (C)
```
```suggestion
    GroundwaterSettings.tempBotm = NaN; % groundwater temperature at bottom layer (C)
```

SarahAlidoost commented:
@BSchilperoort Mostafa and I addressed your comments, please let us know if something else is needed for BMI implemntation. The exe file is re-generated. @MostafaGomaa93 will update the [table](https://github.com/EcoExtreML/STEMMUS_SCOPE/issues/209#issuecomment-2178950772) regarding `topLevel` variable.

MostafaGomaa93 commented:
> @MostafaGomaa93 will update the [table](https://github.com/EcoExtreML/STEMMUS_SCOPE/issues/209#issuecomment-2178950772) regarding `topLevel` variable.

Done

BSchilperoort commented on src/+groundwater/initializeGroundwaterSettings.m:
```diff
@@ -0,0 +1,41 @@
+    GroundwaterSettings.GroundwaterCoupling = 0; % (value = 0 -> deactivate coupling, or = 1 -> activate coupling); default = 0, update value to = 1 -> through BMI
+
+    % Initialize the variables (head, temperature) at the bottom boundary (start of saturated zone)
+    GroundwaterSettings.headBotmLayer = 1950.0; % groundwater head (cm) at bottom layer
```
Why are these arbitrary values here? `1950.0` and `2000.0`?

BSchilperoort commented on src/+groundwater/updateDunnianRunoff.m:
```diff
@@ -0,0 +1,16 @@
+    % (a) direct water input from precipitation when soil is fully saturated (depth to water table = 0)
+    % (b) Return flow (from groundwater exfiltration) calculated in MODFLOW and added to Dunnian runoff (through BMI)
+    % here approach (a) is implemented
+    wat_Dep = groundWaterDepth / 100; % (m);
+
+    dunnianRunoff = zeros(size(Precip_msr));
+    update_Precip_msr = Precip_msr;
+
+    if wat_Dep <= 0.01
```
```suggestion
    dunnianRunoff = zeros(size(Precip_msr));
    update_Precip_msr = Precip_msr;

    if groundWaterDepth <= 1.0
```
No need to convert for a single if statement I think

MostafaGomaa93 replied:
> No need to convert for a single if statement I think

True, the conversion was needed before when this if statement was on another function. But now it's not necessary

BSchilperoort commented on CHANGELOG.md:
```diff
@@ -22,6 +22,7 @@
 - Defining the indices of the first four layers discussed in
   [#237](https://github.com/EcoExtreML/STEMMUS_SCOPE/issues/237) and fixed in
   [#238](https://github.com/EcoExtreML/STEMMUS_SCOPE/pull/238)
+- issue [#240](https://github.com/EcoExtreML/STEMMUS_SCOPE/issues/240)
```
```suggestion
```

I would prefer to leave that issue open until the unused variables are removed.

BSchilperoort reviewed (approved):
Looks good, just some small suggestions.

I think it should be easy to now implement the BMI.

For configuring the groundwater coupling, only the following variables are required, right?

- GroundwaterCoupling
- headBotmLayer
- tempBotm
- topLevel

(aside from the general expansion of the BMI)

MostafaGomaa93 commented:
> For configuring the groundwater coupling, only the following variables are required, right?
> 
> * GroundwaterCoupling
> * headBotmLayer
> * tempBotm
> * topLevel
> 
> (aside from the general expansion of the BMI)

Yes, those are input variables, required in the BMI. However, I mentioned other variables in the [Table](https://github.com/EcoExtreML/STEMMUS_SCOPE/issues/209#issuecomment-2178950772) as outputs (to be passed to MODFLOW)

MostafaGomaa93 commented on src/+groundwater/calculateGroundWaterDepth.m:
```diff
@@ -0,0 +1,13 @@
+    groundWaterDepth = topLevel - headBotmLayer; % depth from top layer to groundwater level
+
+    % Check that the position of the water table is within the soil column
+    if groundWaterDepth <= 0
+        warning('The soil is fully saturated up to the land surface level!');
+        groundWaterDepth = 1.0; % to avoid model crashing, assign minimum groundWaterDepth value of 1 cm
```
> If the minimum depth should be 1 cm, why isn't the if statement checking for that? I.e., `if groundWaterDepth < 1.0`

The minimum depth should be zero, but the code crashes if groundWaterDepth <= 0 and not if groundWaterDepth < 1, so I just set the min = 1 cm, which is the minimum value that the code can use. 

However, values received from MODFLOW can be less than 0 or even negative values

BSchilperoort reviewed (changes requested):
I encountered a bug in the STEMMUS_SCOPE.m file.

The executable file will have to be regenerated. Can you regenerate the file?

SarahAlidoost commented:
> I encountered a bug in the STEMMUS_SCOPE.m file.
> 
> The executable file will have to be regenerated. Can you regenerate the file?

Thanks for fixing the bug, the exe is re-generated.

BSchilperoort commented:
I am able to now use the BMI to set groundwater parameters. It seems to work mostly OK.

However, when not setting the groundwater temperature, the models shows the following behavior:

![image](https://github.com/user-attachments/assets/7b485703-e336-4c1e-a2b5-d0086b94de02)
Y-axis is the groundwater temperature. The thick blue line is the timestep before activating the groundwater coupling, and the other lines the data at subsequent timesteps.

MostafaGomaa93 commented:
> I am able to now use the BMI to set groundwater parameters. It seems to work mostly OK.
> 
Hi @BSchilperoort, I did a quick check with and without setting the groundwater temperature for the ZA-Kru site (1-day simulation). The figure you showed is a bit difficult for me to understand. Could you please check from your side that the model behaviour should be the same as the two figures below? In principle, when groundwater temperature has a value (e.g. case 1 below), then the soil layers below the groundwater table get the same value as the groundwater temperature (in case 1 below -> from thick = 0.5 up to 5 m -> soil temp = groundwater temp = 17).

If the groundwater temperature is Nan, then the temperature of the soil layers below the groundwater table is the same as the temperature of the last soil layer (see case 2 below -> from thick = 0.5 up to 5 m -> soil temp = soil temp of last soil layer, which is = 27.74).

Note, these settings above work from the second time step (not the first one). You can see in both figures, from thick = 0.5 up to 5 m, the temp at the first time step is not the same as in all the following time steps. I don't know why, but its not very important now.

Case 1, groundwater temperature = 17.
![st](https://github.com/user-attachments/assets/22628de8-2b65-4ca9-ab0e-bc96ee694eab)


Case 2, groundwater temperature = Nan.
![st_nan_gwtemp](https://github.com/user-attachments/assets/76e5eb4c-df52-4935-872e-ddccd71ae029)

BSchilperoort commented:
The image I showed contained the temperature vs. depth in the soil.

Your figure contains the same error:
![image](https://github.com/user-attachments/assets/a0806096-780d-4334-ac3d-a68117d95b0e)

MostafaGomaa93 commented:
> The image I showed contained the temperature vs. depth in the soil.
> 
> Your figure contains the same error

I didn't think this is an error. It looks okay

MostafaGomaa93 commented:
> > The image I showed contained the temperature vs. depth in the soil.
> > Your figure contains the same error

 I would like to explain a bit. When we set the temperature of the last soil column (let's call it `Tn`) as a fixed value  (either `Tn` = groundwater temperature or in case of no available groundwater temperature, `Tn` = soil temperature of the soil layer before the bottom layer `Tn-1`), There will be a difference between `Tn` and `Tn-1`. The model is trying to stabilize by minimizing that difference. In the first couple of time steps, that difference is large, but within the time, this difference is smoothed, and that's fine. This behaviour happens for both cases (if groundwater temp is defined or is Nan). so it's not really an error.

BSchilperoort commented:
> In the first couple of time steps, that difference is large, but within the time, this difference is smoothed, and that's fine

I only see this show up when not setting the groundwater temperature. The artifact does not smooth out (at least not after 50 timesteps), and has both a negative and positive component:
![image](https://github.com/user-attachments/assets/fd46957d-0ee3-42f4-a502-aee51692c12e)

BSchilperoort commented:
Additionally, if I don't use the groundwater coupling, but set the soil temperature through the BMI instead, this is the result:
![image](https://github.com/user-attachments/assets/e8ad17f6-5309-458b-aaea-d6cf7cb5ca27)

No artifacts...

MostafaGomaa93 commented:
> **Edit:** I was able to reproduce the above figure by setting the groundwater temperature to 23 deg C. Figure looks the same.

So, you have the same problem if the groundwater temp is set to 23 deg C ?

BSchilperoort commented:
> So, you have the same problem if the groundwater temp is set to 23 deg C ?

Yes. But only though the groundwater coupling interface. Not if I set the soil temperature values directly through the BMI.

MostafaGomaa93 commented:
> > So, you have the same problem if the groundwater temp is set to 23 deg C ?
> 
> Yes. But only though the groundwater coupling interface. Not if I set the soil temperature values directly through the BMI.

We still need to use the groundwater coupling interface. There is something I didn't fully get yet. Perhaps, we could meet online to discuss. Let me know if you have time today or tomorrow

MostafaGomaa93 commented:
I did several tests to track where the problem was. First of all, this strange temperature behaviour happens in the soil layer above the groundwater table and only happens when we set `groundwater temperature` > temperature of the layer above the groundwater table. So, when I was testing with `groundwater temperature = 17 C`, there was no problem. However, when @BSchilperoort changed the value to 23 C, we observed this strange behaviour. The same happens when the `groundwater temperature = NaN` because in this case, the `groundwater temperature` is replaced by the `bottom temperature boundary` which = 22.96 C` (same case as when `groundwater temperature = 23 C`). 

So, in the code, this strange behaviour is related to the `RHS(indxBotm + 1)` which is calculated and updated based on different variables in several lines ([here](https://github.com/EcoExtreML/STEMMUS_SCOPE/blob/a0461c7f4cbe81301f9883a38567bcb1e331a6b2/src/%2Benergy/assembleCoefficientMatrices.m#L72), [here](https://github.com/EcoExtreML/STEMMUS_SCOPE/blob/a0461c7f4cbe81301f9883a38567bcb1e331a6b2/src/%2Benergy/calculateBoundaryConditions.m#L30) and [here](https://github.com/EcoExtreML/STEMMUS_SCOPE/blob/a0461c7f4cbe81301f9883a38567bcb1e331a6b2/src/%2Benergy/solveTridiagonalMatrixEquations.m#L18C1-L25C8)).

Honestly, I tried several things but doesn't solve the problem. Also, the solution suggested by @BSchilperoort doesn't solve it. I came up with an end-up solution to solve the problem (suggested [here](https://github.com/EcoExtreML/STEMMUS_SCOPE/blob/a0461c7f4cbe81301f9883a38567bcb1e331a6b2/src/%2Benergy/solveTridiagonalMatrixEquations.m#L32), not the ideal one so far). A better solution will need a lot of tests and take more time.

From the physics point of view, an important question now is, **Does it happen in reality that groundwater temperature > soil temperature?** I was thinking that this is not the case and that groundwater is always cooler than the soil. But I need to discuss this further with @yijianzeng. I also sent to one of the MODFLOW development team members about this and will update you. 

This is now the temperature figure with `groundwater temperature = 23 C`

![st_gwtemp=23_sol12](https://github.com/user-attachments/assets/94de9491-4364-4a44-bbb9-8361571acd21)

MostafaGomaa93 commented on src/+io/create_output_files_binary.m:
I think we don't need anymore the `gwfluxes` as input to the function

MostafaGomaa93 commented on src/+io/output_data_binary.m:
I think we don't need anymore the `gwfluxes` as input to the function

SarahAlidoost commented on src/+energy/assembleCoefficientMatrices.m:
```diff
@@ -59,11 +59,12 @@
             - (C1(indxBotm, 1) / Delt_t + C4(indxBotm, 1)) * SV.hh(indxBotm) - (C1(indxBotm, 2) / Delt_t + C4(indxBotm, 2)) * SV.hh(indxBotm + 1) ...
             + (C1(indxBotm, 1) / Delt_t) * SV.h(indxBotm) + (C1(indxBotm, 2) / Delt_t) * SV.h(indxBotm + 1);
-        for i = 2:ModelSettings.NL
+        for i = indxBotm:ModelSettings.NL
```
Running the model fails due to this change. In this file, if `~GroundwaterSettings.GroundwaterCoupling` , then `indxBotm=1`, the equation` ARG4 = C1(i - 1, 2) / Delt_t;` returns an error because i-1 is 0. Can you fix it?

MostafaGomaa93 replied:
> Running the model fails due to this change. In this file, if `~GroundwaterSettings.GroundwaterCoupling` , then `indxBotm=1`, the equation` ARG4 = C1(i - 1, 2) / Delt_t;` returns an error because i-1 is 0. Can you fix it?

My mistake, I correct it now

SarahAlidoost commented on src/+soilmoisture/calculateTimeDerivatives.m:
```diff
@@ -47,21 +47,23 @@
             - C6(indxBotm, 1) * P_gg(indxBotm) - C6(indxBotm, 2) * P_gg(2) ...
             + (C2(indxBotm, 1) / Delt_t) * T(indxBotm) + (C2(indxBotm, 2) / Delt_t) * T(indxBotm + 1);
-        for i = 2:ModelSettings.NL
+        for i = indxBotm:ModelSettings.NL
```
Running the model fails due to this change. In this file, if `~GroundwaterSettings.GroundwaterCoupling` , then `indxBotm=1`, the equation` ARG1 = C2(i - 1, 2) / Delt_t;` returns an error because i-1 is 0. Can you fix it?

MostafaGomaa93 replied:
> Running the model fails due to this change. In this file, if `~GroundwaterSettings.GroundwaterCoupling` , then `indxBotm=1`, the equation` ARG1 = C2(i - 1, 2) / Delt_t;` returns an error because i-1 is 0. Can you fix it?

My mistake, I correct it now

SarahAlidoost commented on src/+groundwater/calculateGroundwaterRecharge.m:
```diff
@@ -64,18 +64,21 @@ soilThick               cumulative soil layers thickness (from top to bottom)
         recharge_init = Q_flip(indxRchrg); % mean([Q_flip(indxRchrg), Q_flip(indxRchrg_above)])
     end
 
+    %{
```
@MostafaGomaa93 can you please address this comment? The block `%{%}` is intended for documentation after the function name and not for commenting out the codes.

SarahAlidoost commented:
@MostafaGomaa93 can you please address the comments above? Then I can create the exe file and test whether the model runs with/without coupling.

MostafaGomaa93 commented on src/STEMMUS_SCOPE.m:
```diff
@@ -777,6 +777,7 @@
                                       Evap, WaterStressFactor, WaterPotential, Sim_hh, Sim_qlh, Sim_qlt, Sim_qvh, ...
                                       Sim_qvt, Sim_qla, Sim_qva, Sim_qtot, ForcingData, RS, RWUs, RWUg, GroundwaterSettings, gwfluxes);
         fclose("all");
+        disp(['Solving Time step: ' num2str(KT) ' out of ' num2str(endTime)]); 
```
> For now, please remove the `disp` message.

Removed

MostafaGomaa93 commented:
> @MostafaGomaa93 can you please address the comments above? Then I can create the exe file and test whether the model runs with/without coupling.

Done

BSchilperoort commented:
I'm afraid this did not fully fix the issue... While there is no negative spike anymore, a growing positive spike does form over time:

![image](https://github.com/user-attachments/assets/17afa6ae-54bf-445f-ab57-1160337c26bd)

SarahAlidoost commented:
> I'm afraid this did not fully fix the issue... While there is no negative spike anymore, a growing positive spike does form over time:
> 
> ![image](https://private-user-images.githubusercontent.com/12114825/350850027-17afa6ae-54bf-445f-ab57-1160337c26bd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE2MzcyNDIsIm5iZiI6MTcyMTYzNjk0MiwicGF0aCI6Ii8xMjExNDgyNS8zNTA4NTAwMjctMTdhZmE2YWUtNTRiZi00NDVmLWFiNTctMTE2MDMzN2MyNmJkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzIyVDA4MjkwMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM3MTU3Yjg4YTcwMzQ1YTUxZTBjMDYxYzM1ZTI0MjU2NDU0ZTgxMmRmNTUzOWE3ZGNlMzk2ODdhZGRlYzkyYWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.RsclrEzuqQLeiCasFIr2nTEBo5FrA2cELDX7QLkAlpU)

@BSchilperoort and @MostafaGomaa93 It seems that fixing this issue of temperature needs more work and tests and this pull request has expanded beyond just addressing unused variables. Since the BMI functionality is now working, I prefer to merge this pull request and the BMI one in Pystemmusscope. The other issue should be resolved in a separate pull request.

BSchilperoort commented:
> I prefer to merge this pull request and the BMI one in Pystemmusscope. The other issue should be resolved in a separate pull request.

We can merge this PR, however, I don't think it's a good idea to make a release yet until the issue has been fixed.

We cannot merge a working PR on PyStemmusScope without there being a new STEMMUS_SCOPE container image (which ideally requires a new release).

yijianzeng commented:
> From the physics point of view, an important question now is, **Does it happen in reality that groundwater temperature > soil temperature?** I was thinking that this is not the case and that groundwater is always cooler than the soil. But I need to discuss this further with @yijianzeng

Dear all, thanks a lot for your great efforts into this effort. Indeed, physically speaking for the soil bottom when it is emerged in groundwater, the soil temperature should be equal to the groundwater temperature. As such, if you have an initial soil temperature like 17-19 degree or so, the groundwater temperature should be set equal to that value. 

To consider the heat flow of groundwater is indeed another topic to tackle with, since this is linked to MODFLOW's heat flow version, which is not yet released yet as i understood.  

For now, I would suggest fixing the groundwater temperature = soil temperature wherever GW reaches a certain depth of the soil column.
